pimd/pim_register.c

```diff
@@ -131,32 +152,68 @@ int pim_register_stop_recv(struct interface *ifp, uint8_t *buf, int buf_size)
 	pim_parse_addr_ucast(&source, buf, buf_size);
 	sg.src = source.u.prefix4;
 
+	rp = RP(pim_ifp->pim, sg.grp);
+
+	if (sg.src.s_addr == rp->rpf_addr.u.prefix4.s_addr)
+		sg.src.s_addr = INADDR_ANY;
+
 	upstream = pim_upstream_find(pim, &sg);
+
+	if (PIM_DEBUG_PIM_REG) {
+		if (upstream)
+			zlog_debug("Received Register stop for %s",
+				   upstream->sg_str);
+		else
+			zlog_debug(
+				"Received Register stop for %s : no upstream",
+				pim_str_sg_dump(&sg));
+	}
+
+	if ((!upstream) && (sg.src.s_addr != INADDR_ANY))
+		return 0;
+
+	/*
+	 * RFC 7761 Sec 4.4.1
+	 * Handling Register-Stop(*,G) Messages at the DR:
+	 *   A Register-Stop(*,G) should be treated as a Register-Stop(S,G) for
+	 * all (S,G) Register state machines that are not in the NoInfo state.
+	 */
+
+	/* No starg upstream. loop upstream list to get all sg */
 	if (!upstream) {
+		struct pim_upstream *up;
```

Comment:
Let's simplify this further and get rid of the basically identical two loops.  If we do not have an upstream do a pim_upstream_find for the *,G and then if nothing there just return else fall through and use the upstream from either source to loop one time.  This loop is terribly inefficient the way it is written and will be awful at scale.

Reply:
I do not agree with performance degradation.

Please note that loops are not identical. One loop on upstream list and other was on children of upstream.
if starg exists, we are going to run only on children.
if starg doesn't exist we will scan entire.
There is no way loop executes twice.

Anyways I have changed the code to look bit more simpler.

Reply:
it is a serious performance degradation.  When you have 100's of k of upstream looking at each one makes no fucking sense.  Especially since we already know the list of all S,G's for a *,G.

Here's the code:

if (sg.src.s_addr == rp->rpf_addr.u.prefix4.s_addr)
     sg.src.s_addr = INADDR_ANY;

up = pim_upstream_find(pim, &sg);

if (!up)
     return;

pim_reg_stop_upstream(up);

for (ALL_LIST_ELEMENTS_RO(up->sources, up_node, child)) {
     if (pim_debug)
         the log

   pim_reg_stop_upstream(child);
}

Reply:
What about those S,G which do not have *,G parent. Why do you assume S,G always have *,G parent?
In that scenario there is no other go but to scan entire entries.